bugsnag-android-core/src/main/java/com/bugsnag/android/Configuration.java

```diff
@@ -23,11 +23,8 @@
  * User-specified configuration storage object, contains information
  * specified at the client level, api-key and endpoint configuration.
  */
-public class Configuration extends Observable implements Observer {
```

Comment:
I'm concerned that this change might introduce errors for people upgrading if they have code that looks like

```
val config = Configuration("my-api-key")
config.endpoints = Endpoints("http://localhost:8000", "http://localhost:10000")
Bugsnag.init(this, config)

config.releaseStage = "test" // Will no longer apply the release stage to the running bugsnag
```
Then it will start to fail silently after the change

I wonder if it would be better to have a builder pattern for the new configuration, where the immutable config is only available as part of the builder:

```
val configBuilder = ConfigurationBuilder("my-api-key")
configBuilder.endpoints = Endpoints("http://localhost:8000", "http://localhost:10000")
val config = configBuilder.build()
Bugsnag.init(this, config)

config.releaseStage = "test"  // Error cannot set releaseStage
```

That would mean that errors would be discovered at compile time

Reply:
I think you make a good point. With the builder interface would that be a compile or a runtime error? If runtime you could do implement this with the current interface – `Configuration` has a `private boolean consumed` field which gets set to `true` when passed to `Bugsnag.init()` and subsequent setters check `consumed==false` and throw otherwise.

If this would be a compile error, arguably that's much better.

Reply:
The way I've written that example I think it would be a compile time error (but that is a bit contrived). In reality it would probably be written more like:

```
Bugsnag.init(this, new ConfigBuilder()
            .setEndpoints("http://localhost:8000", "http://localhost:10000")
            .setReleaseStage("test")
            .build()
)
```
which means that the config object can't be accessed directly after bugsnag has been initialised anyway

Reply:
I'm not sure the builder route makes as much sense on Kotlin as it would Java, since the convention there is to use named parameters with default values.

We could certainly throw an exception from all setters once `Configuration` has been consumed. We'd have to be sure that this would be worth the complexity/maintenance overhead for whenever a new property is added

Reply:
What does the java interface look like if you use named and default parameters in a kotlin class? The builder pattern in kotlin isn't that bad if it makes the java side much better - but we really want to pick the lesser of two evils here I think.

Reply:
Would offering both make sense? We could rely on proguard to strip the builder class if someone is using kotlin and not referencing the builder? I don't know if thats a sensible/possible suggestion - just throwing out ideas to try and get the best of both rather than the lesser of two evils.